### PR TITLE
REP Complete! (4-2-stable)

### PR DESCRIPTION
**Pull request by korydraughn**
Ready for release.

### CODE REVIEW

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 1]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
-
-#       self.run_test(f, self.admin.session_collection, max_number_of_objects, max_size_in_bytes)
-
-
     @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
-    def test_copy_and_remove_data_object(self):
-        max_number_of_objects = 100
-        max_size_in_bytes = 10000
+    def test_control_rules(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 2]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
+            expected_size_in_bytes = 0
+            self.assert_quotas(sandbox, expected_number_of_objects, expected_size_in_bytes)
 
-                # Check quotas.
-                values = self.get_logical_quotas_attribute_values(self.admin.session_collection)
-                self.assertEquals(values[self.current_number_of_objects_attribute()], expected_number_of_objects)
-                self.assertEquals(values[self.current_size_in_bytes_attribute()],     expected_size)
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_incorrect_config(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 3]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
+            config.server_config['plugin_configuration']['rule_engines'].insert(0, {
+                'instance_name': 'irods_rule_engine_plugin-logical_quotas-instance',
+                'plugin_name': 'irods_rule_engine_plugin-logical_quotas',
+                'plugin_specific_configuration': {
+                    'namespace': self.logical_quotas_namespace(),
+                    'metadata_attribute_names': {
+                        'mum_numberata_objects': self.maximum_number_of_data_objects_attribute_name(),
+                        'aximum_': self.maximum_size_in_bytes_attribute_name(),
+                        'umber_of_data': self.total_number_of_data_objects_attribute_name(),
+                        'total_size_in': self.total_size_in_bytes_attribute_name()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: these names don't seem... correct

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 4]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
+            self.admin.assert_icommand_fail(['iput', filename], 'STDOUT', [error_msg])
+            os.remove(filename)
 
-            for filename in file_copies:
-                source = filename['source']
-                target = filename['target']
-                size = filename['size']
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_put_data_object(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 5]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
-                self.assertEquals(values[self.current_number_of_objects_attribute()], expected_number_of_objects)
-                self.assertEquals(values[self.current_size_in_bytes_attribute()],     expected_size)
+            self.logical_quotas_stop_monitoring_collection(sandbox)
 
-            for filename in filenames:
-                name = filename['name']
-                size = filename['size']
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_put_collection(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 6]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
-        self.run_test(f, self.admin.session_collection, max_number_of_objects, max_size_in_bytes)
+            self.logical_quotas_stop_monitoring_collection(sandbox)
 
-#   @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
-#   def test_copy_and_remove_collection(self):
-#       max_number_of_objects = 3
-#       max_size_in_bytes = 3000
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_copy_data_object(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 7]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
-            self.assertEquals(values[self.current_size_in_bytes_attribute()],     1024)
-
-            # Remove the recently added data object and check that the quota values for
-            # the collection decrease appropriately.
-            self.admin.assert_icommand(['irm', '-f', filename])
-            values = self.get_logical_quotas_attribute_values(self.admin.session_collection)
-            self.assertEquals(values[self.current_number_of_objects_attribute()], 0)
-            self.assertEquals(values[self.current_size_in_bytes_attribute()],     0)
+    def test_copy_collection(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 8]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
+            expected_size_in_bytes += file_size
+            self.assert_quotas(col2, expected_number_of_objects, expected_size_in_bytes)
 
-#   @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
-#   def test_put_and_remove_collection(self):
-#       max_number_of_objects = 3
-#       max_size_in_bytes = 3000
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_rename_data_object(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 9]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
+            self.assert_quotas(col2, 0, 0)
+            self.assert_quotas(col3, expected_number_of_objects, expected_size_in_bytes)
 
-#   @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
-#   def test_rename_and_remove_data_object(self):
-#       max_number_of_objects = 3
-#       max_size_in_bytes = 3000
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_rename_collection(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: packaging/test_rule_engine_plugin_logical_quotas.py [hunk 10]
```diff
@@ -25,208 +26,553 @@ def setUp(self):
...
+            self.assert_quotas(col1, expected_number_of_objects, expected_size_in_bytes)
+            self.assert_quotas(col2, 0, 0)
 
-#   @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
-#   def test_rename_and_remove_collection(self):
-#       max_number_of_objects = 3
-#       max_size_in_bytes = 3000
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "Skip for Topology Testing")
+    def test_stream_data_object(self):
+	config = IrodsConfig()
```
Comment (trel) [inline packaging/test_rule_engine_plugin_logical_quotas.py]: whitespace/indentation

File: src/handler.cpp [hunk 11]
```diff
@@ -995,145 +1079,331 @@ namespace handler {
...
+            const auto& attrs = get_instance_config(_instance_configs, _instance_name).attributes();
+            const auto& l1desc = irods::get_l1desc(input->l1descInx); 
+
+            // TODO Does opening a data object with the O_TRUNC flag cause an update to the catalog?
+            // This is important because for lseek, we need the real size of the data object. In this case,
+            // zero should be returned.
+            //
+            // TODO Updating the catalog to reflect O_TRUNC is mandatory and should happen in 4.2.8.
+            // 
+            // For now, we can use dstream::seekg/tellg to retreive the real size of the data object
```
Comment (trel) [inline src/handler.cpp]: retrieve

File: src/handler.hpp [hunk 12]
```diff
@@ -71,15 +73,27 @@ namespace handler {
...
+
+        static auto reset() noexcept -> void;
 
-    auto pep_api_data_obj_copy_post(const std::string& _instance_name,
-                                    const instance_configuration_map& _instance_configs,
-                                    std::list<boost::any>& _rule_arguments,
-                                    irods::callback& _effect_handler) -> irods::error;
+        static auto pre(const std::string& _instance_name,
+                        const instance_configuration_map& _instance_configs,
+                        std::list<boost::any>& _rule_arguments,
```
Comment (trel) [inline src/handler.hpp]: std::any ?

File: src/main.cpp [hunk 13]
```diff
@@ -102,25 +105,47 @@ namespace
...
-                                                                   attr_names.at("maximum_size_in_bytes").get<std::string>(),
-                                                                   attr_names.at("total_number_of_data_objects").get<std::string>(),
-                                                                   attr_names.at("total_size_in_bytes").get<std::string>()}};
+                    const auto& attr_names = [&plugin_config] {
+                        try {
+                            return plugin_config.at("metadata_attribute_names");
+                        }
+                        catch (...) {
+                            throw std::runtime_error{fmt::format("Logical Quotas Policy: Failed to find rule engine "
+                                                                 "plugin configuration property [metadata_attribute_name]")};
```
Comment (trel) [inline src/main.cpp]: metadata_attribute_name -> metadata_attribute_names

File: src/main.cpp [hunk 14]
```diff
@@ -157,22 +182,20 @@ namespace
...
     {
-        const auto iter = pep_handlers.find(_rule_name);
-        
-        if (iter != std::end(pep_handlers)) {
+        if (const auto iter = pep_handlers.find(_rule_name); iter != std::end(pep_handlers)) {
             return (iter->second)(_instance_name, instance_configs, _rule_arguments, _effect_handler);
         }
 
-        rodsLog(LOG_ERROR, "Rule not supported in rule engine plugin [rule => %s]", _rule_name.c_str());
+        rodsLog(LOG_ERROR, "[logical_quota] Rule not supported in rule engine plugin [rule => %s]", _rule_name.c_str());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 15]
```diff
@@ -102,25 +105,47 @@ namespace
...
+
+                    instance_configs.insert_or_assign(_instance_name, instance_config);
 
                     return SUCCESS();
                 }
             }
         }
         catch (const std::exception& e) {
-            rodsLog(LOG_ERROR, e.what());
+            rodsLog(LOG_ERROR, "[logical_quota] Bad rule engine plugin configuration");
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 16]
```diff
@@ -157,22 +182,20 @@ namespace
...
         return CODE(RULE_ENGINE_CONTINUE);
     }
 
     auto exec_rule_text_impl(const std::string& _instance_name,
-                             std::string _rule_text,
+                             std::string_view _rule_text,
                              irods::callback _effect_handler) -> irods::error
     {
-        rodsLog(LOG_DEBUG, "_rule_text (before) = %s", _rule_text.c_str());
+        rodsLog(LOG_DEBUG, "[logical_quota] _rule_text => %s", _rule_text.data());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 17]
```diff
@@ -185,17 +208,16 @@ namespace
...
         }
 
-        rodsLog(LOG_DEBUG, "_rule_text (after) = %s", _rule_text.c_str());
+        rodsLog(LOG_DEBUG, "[logical_quota] _rule_text => %s", std::string{_rule_text}.data());
 
         try {
             const auto json_args = json::parse(_rule_text);
 
-            rodsLog(LOG_DEBUG, "json_arguments => %s", json_args.dump().c_str());
+            rodsLog(LOG_DEBUG, "[logical_quota] json_arguments => %s", json_args.dump().data());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 18]
```diff
@@ -185,17 +208,16 @@ namespace
             _rule_text = _rule_text.substr(start, _rule_text.rfind(" }") - start);
         }
 
-        rodsLog(LOG_DEBUG, "_rule_text (after) = %s", _rule_text.c_str());
+        rodsLog(LOG_DEBUG, "[logical_quota] _rule_text => %s", std::string{_rule_text}.data());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 19]
```diff
@@ -210,19 +232,19 @@ namespace
             return ERROR(INVALID_OPERATION, fmt::format("Invalid operation [{}]", op));
         }
         catch (const json::parse_error& e) {
-            rodsLog(LOG_ERROR, e.what());
+            rodsLog(LOG_ERROR, "[logical_quota] error => %s", e.what());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 20]
```diff
@@ -210,19 +232,19 @@ namespace
...
             return ERROR(USER_INPUT_FORMAT_ERR, e.what());
         }
         catch (const json::type_error& e) {
-            rodsLog(LOG_ERROR, e.what());
+            rodsLog(LOG_ERROR, "[logical_quota] error => %s", e.what());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 21]
```diff
@@ -210,19 +232,19 @@ namespace
...
             return ERROR(SYS_INTERNAL_ERR, e.what());
         }
         catch (const std::exception& e) {
-            rodsLog(LOG_ERROR, e.what());
+            rodsLog(LOG_ERROR, "[logical_quota] error => %s", e.what());
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: src/main.cpp [hunk 22]
```diff
@@ -210,19 +232,19 @@ namespace
...
             return ERROR(SYS_INTERNAL_ERR, e.what());
         }
         catch (...) {
-            rodsLog(LOG_ERROR, "Unknown error");
+            rodsLog(LOG_ERROR, "[logical_quota] Unknown error");
```
Comment (trel) [inline src/main.cpp]: logical_quota -> logical_quotas

File: packaging/test_rule_engine_plugin_logical_quotas.py (hunk 3 above)
Comment (korydraughn) [reply packaging/test_rule_engine_plugin_logical_quotas.py]: That test is verifying handling of bad configs.

Comment (trel) [reply packaging/test_rule_engine_plugin_logical_quotas.py]: ha, i knew that.

File: packaging/test_rule_engine_plugin_logical_quotas.py (hunk 1 above)
Comment (korydraughn) [reply packaging/test_rule_engine_plugin_logical_quotas.py]: I'm not sure why the indentation is appearing like that.  the indentation is correct in my editor.  I'll look into that.

Comment (korydraughn) [reply packaging/test_rule_engine_plugin_logical_quotas.py]: Fixed it.

File: src/handler.hpp (hunk 12 above)
Comment (korydraughn) [reply src/handler.hpp]: That is part of the iRODS interface. We could possibly change it to std::any, but I think we'd have to change it everywhere then.

Comment (trel) [reply src/handler.hpp]: agreed.   just realizing with the nice things... we could phase out the boostisms

**Pull request merged**